### DISCUSSION TITLE
New Avalonia Project Missing "iOS Local Device" As Debug Target

### DISCUSSION THREAD

**Discussion by JGKle** [AvaloniaUI/Avalonia#14220] [Q&A] [answered]
I'm trying to get iOS Hot Restart working - per the documentation https://learn.microsoft.com/en-us/xamarin/xamarin-forms/deploy-test/hot-restart , and as shown on various YouTube tutorials (i.e. https://www.youtube.com/watch?v=snQ1C6Cppr8&t=10m20s), Hot Restart allows you to debug on a locally tethered iOS device.

On MAUI, it works right out of the box: when you create a new project, the "Debug" dropdown menu contains the expected "iOS Local Devices" item:

![2024-01-15 20 57 04](https://github.com/AvaloniaUI/Avalonia/assets/1877409/b1b7eca6-c5a4-4ada-91f9-85015fff641a)

When selected, it brings up the Setup Hot Restart wizard. However, in a new Avalonia project, there's never a "Local Devices" option (even with its AvaloniaProject.iOS project set as the startup project):

![2024-01-15 20 58 18](https://github.com/AvaloniaUI/Avalonia/assets/1877409/90397749-b246-4cec-b8c2-23e3277e7e66)

How can I get Hot Restart going in Avalonia, as in MAUI?

**Comment by maxkatz6** [comment, marked as answer]
Hot Restart is a VS feature hardcoded to the MAUI. It's not something we can integrate into Avalonia.

**Reply by JGKle** [reply]
Aw, bummer. So I assume that means there's no way to directly debug on an iOS device from Windows at all with Avalonia?

**Reply by maxkatz6** [reply]
Only with macOS

**Reply by JGKle** [reply]
Dang :(  Thanks for the replies

**Reply by JGKle** [reply]
Follow-up: It seems like https://github.com/AvaloniaUI/AvaloniaMauiHybrid provides some sort of combinations of Maui+Avalonia. There's nothing hacky that could be done to the iOS project only, so I could get HotReload to work? Or like have the iOS app actually be MAUI but then use the same Avalonia views+viewModels as the other apps, which actually use Avalonia completely?

**Reply by maxkatz6** [reply]
Yes, MAUI hybrid should work this way. The only really "hacky" part is that you would need to ship app with both frameworks, potentially increasing app size (and headache maybe as well).

**Reply by JGKle** [reply]
Great! Yeah, Apple's absurd insistence on routing everything through a Mac is just way too much of a hassle. Even when working at home, but especially when traveling, as then you either have to lug around 2 computers or connect to a Mac remotely over the Internet. Would be much easier to be able to just plug an iPhone into my PC directly.

So then, given an existing Avalonia app, what would be the best way to make this happen? i.e. do I need to start a whole new Maui app, add it to my solution, get rid of the Avalonia.iOS project, & follow the `Embedding Avalonia Controls Inside MAUI` approach documented in AvaloniaMauiHybrid? Or can I just add `<UseMaui/>` to my existing Avalonia project & presumably another thing or 2 to tell VS that it can HotReload?

**Reply by JGKle** [reply]
So fyi, it looks like even the sample/demo project of embedding Maui in an Avalonia app doesn't work: https://github.com/AvaloniaUI/AvaloniaMauiHybrid/tree/main/AvaloniaSample. I posted about it here: https://github.com/AvaloniaUI/AvaloniaMauiHybrid/issues/23

**Reply by MikeCodesDotNET** [reply]
> So fyi, it looks like even the sample/demo project of embedding Maui in an Avalonia app doesn't work: https://github.com/AvaloniaUI/AvaloniaMauiHybrid/tree/main/AvaloniaSample. I posted about it here: https://github.com/AvaloniaUI/AvaloniaMauiHybrid/issues/23

As @maxkatz6 has already said, this isn't a supported feature. It's always been sensible to use a Mac if you're developing for the Apple ecosystem.

**Reply by JGKle** [reply]
> As @maxkatz6 has already said, this isn't a supported feature

What? I'm reporting that the demo project in that repo doesn't run as-is - Max definitely didn't say AvaloniaMauiHybrid isn't a supported feature...

Or are you saying that within AvaloniaMauiHybrid, the MauiSample project is expected to work from Windows but the AvaloniaSample.iOS project is not?

(If that's the case...that's pretty much what I was trying to clarify in my question on [June 23](https://github.com/AvaloniaUI/Avalonia/discussions/14220#discussioncomment-9854082). Max said that HotReload should work with both Maui & Avalonia, and both of these projects have both Maui & Avalonia. So if that's what you're saying - which I'm still not sure of - it certainly isn't clear that one Maui+Avalonia combination should work and another Maui+Avalonia combination shouldn't)

**Reply by maxkatz6** [reply]
@metal450 
It's not supported in normal Avalonia. MAUI Hybrid is another story.
Let's keep `MAUI Hybrid` issue discussion in its repository.

**Reply by JGKle** [reply]
> It's not supported in normal Avalonia. MAUI Hybrid is another story.

Right, I was just responding to ur last comment - "MAUI hybrid should work this way". Since "MAUI hybrid" can mean 2 different things, was inquiring if you meant either would work, or only one.

> Let's keep MAUI Hybrid issue discussion in its repository.

Ok yeah, maybe I shouldn't have cross-posted the issue I was seeing here. It was just in the context of the prior discussion tho - so it seemed to make more sense to say "...of the 2 approaches, it seems like only one is working", rather than starting the thread over from scratch over there to re-ask which of the 2 Hybrid approaches u meant could be used for this purpose...

**Comment by MikeCodesDotNET** [comment]
> Hot Restart is a VS feature hardcoded to the MAUI. It's not something we can integrate into Avalonia.

Unless the application is a MAUI app, the Hot Reload feature will not work. 

The MAUI Hybrid projects contain two approaches: 

1. Embedding Avalonia in MAUI (MAUI Sample)
2. Embedding MAUI in Avalonia (Avalonia Sample). 

If any one of these is possibly going to work (not guaranteed), it'd be the MAUI Sample. My recommendation would be to always use a Mac for iOS development though, as the Visual Studio tooling for iOS development has always been pretty unreliable, even back in the days of Xamarin.

**Reply by JGKle** [reply]
> If any one of these is possibly going to work (not guaranteed), it'd be the MAUI Sample.

It looks like that one's working. I have ~80% of the app running already, just some kinks to work out :)

> My recommendation would be to always use a Mac for iOS development though

Yeah, the thing is, I work remotely most of the time, & it isn't really feasible to lug around an entire 2nd laptop just because Apple wants to force everyone to buy more hardware. I did have to get a basic Mac Mini so they'd grant me the privilege of building for the AppStore, but for day-to-day development, it would be far easier to use the same system as I use for everything else. (Also, a Mac with similar performance would cost over double, can never be upgraded, and can't be repaired unless I happen to be close enough to an "Apple Approved" shop). So to me it's just the lesser of evils (if I can make it work of course).

In practice I'm able to do the vast majority of the development on Desktop or Android anyway, it's just the last iOS-specific odds & ends that I have to actually test on Apple. If I can use Maui as a wrapper to be able to test that via direct tethering, then I can avoid having to spend thousands more & lug around two laptops instead of one.
